src/rule.rs

```diff
@@ -662,35 +613,33 @@ pub trait Folder<'cx, Pat: 'cx + Eq + Hash>: Sized {
     }
 }
 
-impl RuleWithNamedFields {
-    // HACK(eddyb) this is pretty expensive, find a better way
-    fn filter_fields<'a>(&'a self, i: Option<usize>) -> impl Iterator<Item = (IStr, Field)> + 'a {
-        self.fields.iter().filter_map(move |(&name, field)| {
-            let paths: IndexMap<_, _> = field
-                .paths
-                .iter()
-                .filter_map(move |(path, subfields)| {
-                    if path.first().cloned() == i {
-                        Some((path.get(1..).unwrap_or(&[]).to_vec(), subfields.clone()))
-                    } else {
-                        None
-                    }
-                })
-                .collect();
-            if !paths.is_empty() {
-                Some((name, Field { paths }))
-            } else {
-                None
-            }
-        })
-    }
-
+impl RuleWithFields {
     pub fn fold<'cx, Pat: 'cx + Eq + Hash>(self, folder: &mut impl Folder<'cx, Pat>) -> Self {
-        let field_rule = |rule, i| RuleWithNamedFields {
+        let cx = folder.cx();
+        let aggregate_fields = match cx[self.fields] {
+            Fields::Leaf(Some(field)) => {
+                let mut rule = RuleWithFields {
+                    rule: self.rule,
+                    fields: field.sub,
+                }
+                .fold(folder);
+                rule.fields = cx.intern(Fields::Leaf(Some(Field {
+                    name: field.name,
+                    sub: rule.fields,
+                })));
+                return rule;
+            }
+            Fields::Leaf(None) => &[][..],
+            Fields::Aggregate(ref children) => children,
+        };
+        let field_rule = |rule, i| RuleWithFields {
             rule,
-            fields: self.filter_fields(Some(i)).collect(),
+            fields: aggregate_fields
```

Comment:
This closure could be named `child` and `aggregate_fields` could be named `children`.